src/iproute2_sysrepo.c

```diff
@@ -116,15 +143,182 @@ static const struct cmd {
+}
+
+static int print_nofopt(struct filter_util *qu, FILE *f, struct rtattr *opt, __u32 fhandle)
+{
+	if (opt && RTA_PAYLOAD(opt))
+		fprintf(f, "fh %08x [Unknown filter, optlen=%u] ",
+			fhandle, (unsigned int) RTA_PAYLOAD(opt));
+	else if (fhandle)
+		fprintf(f, "fh %08x ", fhandle);
+	return 0;
+}
+
+static int parse_nofopt(struct filter_util *qu, char *fhandle,
+			int argc, char **argv, struct nlmsghdr *n)
+{
+	__u32 handle;
+
+	if (argc) {
+		fprintf(stderr,
+			"Unknown filter \"%s\", hence option \"%s\" is unparsable\n",
+			qu->id, *argv);
+		return -1;
+	}
+	if (fhandle) {
+		struct tcmsg *t = NLMSG_DATA(n);
+
+		if (get_u32(&handle, fhandle, 16)) {
+			fprintf(stderr, "Unparsable filter ID \"%s\"\n", fhandle);
+			return -1;
+		}
+		t->tcm_handle = handle;
+	}
+	return 0;
+}
+
+struct qdisc_util *get_qdisc_kind(const char *str)
+{
+	void *dlh;
+	char buf[256];
+	struct qdisc_util *q;
+
+	for (q = qdisc_list; q; q = q->next)
+		if (strcmp(q->id, str) == 0)
+			return q;
+
+	snprintf(buf, sizeof(buf), "%s/q_%s.so", get_tc_lib(), str);
+	dlh = dlopen(buf, RTLD_LAZY);
+	if (!dlh) {
+		/* look in current binary, only open once */
+		dlh = BODY;
+		if (dlh == NULL) {
+			dlh = BODY = dlopen(NULL, RTLD_LAZY);
+			if (dlh == NULL)
+				goto noexist;
+		}
+	}
+
+	snprintf(buf, sizeof(buf), "%s_qdisc_util", str);
+	q = dlsym(dlh, buf);
+	if (q == NULL)
+		goto noexist;
+
+reg:
+	q->next = qdisc_list;
+	qdisc_list = q;
+	return q;
+
+noexist:
+	q = calloc(1, sizeof(*q));
+	if (q) {
+		q->id = strdup(str);
+		q->parse_qopt = parse_noqopt;
+		q->print_qopt = print_noqopt;
+		goto reg;
+	}
+	return q;
+}
+
+
+struct filter_util *get_filter_kind(const char *str)
+{
+	void *dlh;
+	char buf[256];
+	struct filter_util *q;
+
+	for (q = filter_list; q; q = q->next)
+		if (strcmp(q->id, str) == 0)
+			return q;
+
+	snprintf(buf, sizeof(buf), "%s/f_%s.so", get_tc_lib(), str);
+	dlh = dlopen(buf, RTLD_LAZY);
+	if (dlh == NULL) {
+		dlh = BODY;
+		if (dlh == NULL) {
+			dlh = BODY = dlopen(NULL, RTLD_LAZY);
+			if (dlh == NULL)
+				goto noexist;
+		}
+	}
+
+	snprintf(buf, sizeof(buf), "%s_filter_util", str);
+	q = dlsym(dlh, buf);
+	if (q == NULL)
+		goto noexist;
+
+reg:
+	q->next = filter_list;
+	filter_list = q;
+	return q;
+noexist:
+	q = calloc(1, sizeof(*q));
+	if (q) {
+		strncpy(q->id, str, 15);
+		q->parse_fopt = parse_nofopt;
+		q->print_fopt = print_nofopt;
+		goto reg;
+	}
+	return q;
+}
+
 static int do_cmd(const char *argv0, int argc, char **argv)
 {
+    const struct cmd *cmds;
     const struct cmd *c;
+    if (!strcmp(argv0,"ip"))
+        cmds = ip_cmds;
+    else if (!strcmp(argv0,"bridge"))
+        cmds = bridge_cmds;
+    else if (!strcmp(argv0,"tc"))
+        cmds = tc_cmds;
+    else {
+        fprintf(stderr, "Command '%s' is unknown\n",argv0);
+        return EXIT_FAILURE;
+    }
 
-    for (c = cmds; c->cmd; ++c) {
-        if (matches(argv0, c->cmd) == 0)
-            return -(c->func(argc - 1, argv + 1));
+    if (argc > 1){
```

Comment:
`do_cmd` should take `agrv0` (ip, tc, bridge), and `argv1` (link, route, .... etc), the argc validation should be at `main`

something like this:
```c
static int do_cmd(const char* argv0 ,const char *argv1, int argc, char **argv)
{
    struct cmd **cmds;
    if (!strcmp(argv0,"ip")
         cmds= ip_cmds;
     if (!strcmp(argv0,"bridge")
         cmds= brdige_cmds;
     if (!strcmp(argv0,"tc")
         cmds= tc_cmds;
    
    const struct cmd *c;
    for (c = cmds; c->cmd; ++c) {
        if (matches(argv1, c->cmd) == 0)
            return -(c->func(argc - 2, argv + 2));
    }
    return EXIT_FAILURE;
}
```